### PR TITLE
Deployment preparation: L2 gauge checkpointer

### PR DESCRIPTION
**Pull request by jubeira**
Adding fork tests and deployment artifacts / scripts for L2 gauge checkpointer.

Fork test block # was set close to the time where the fork test was last updated.

Build info from https://github.com/balancer-labs/balancer-v2-monorepo/pull/1601/commits/db9ab19541945ed8419cdf2f716b22e8b5f31bd8.

Note that at https://github.com/balancer-labs/balancer-v2-monorepo/pull/1601/commits/1895091f1dab5d4b67f36d72fa44edc05309b92a the updates from master already have the changes from the entrypoint, so the build info generated at that point will be different and incompatible with the fork test.

### CODE REVIEW

File: pkg/deployments/README.md [hunk 1]
```diff
@@ -102,6 +102,7 @@ Returns an object with all contracts from a deployment and their addresses.
 | Batch Relayer V4                                        | [`20220916-batch-relayer-v4`](./tasks/20220916-batch-relayer-v4)                                     |
 | Managed Pool                                            | [`20221021-managed-pool`](./tasks/20221021-managed-pool)                                             |
 | Protocol Fee Splitter                                   | [`20221027-fee-splitter`](./tasks/20221027-fee-splitter)                                             |
+| L2 Gauge Checkpointer                                   | [`20221101-l2-gauge-checkpointer`](./tasks/20221101-l2-gauge-checkpointer)                           |
```
Comment (nventuro) [inline pkg/deployments/README.md:105]: Can you also update the changelog?

Comment (nventuro) [reply pkg/deployments/README.md:105]: Both of deployments and interfaces (which should have been done in #1570)

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+import { GaugeType } from '@balancer-labs/balancer-js/src/types';
+import * as expectEvent from '@balancer-labs/v2-helpers/src/test/expectEvent';
+
+import { describeForkTest } from '../../../src/forkTests';
+import Task, { TaskMode } from '../../../src/task';
+import { getForkedNetwork } from '../../../src/test';
+import { impersonate } from '../../../src/signers';
+
+// This block number is before the manual weekly checkpoint. This ensures gauges will actually be checkpointed.
+describeForkTest('L2GaugeCheckpointer', 'mainnet', 15272610, function () {
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:15]:
```suggestion
describeForkTest('L2GaugeCheckpointer', 'mainnet', 15835800, function () {
```

Let's do a more recent one

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    ['0x899F737750db562b88c1E412eE1902980D3a4844', 1],
+    ['0x6cb1A77AB2e54d4560fda893E9c738ad770da0B0', 1],
+    ['0x6823DcA6D70061F2AE2AAA21661795A2294812bF', 1],
+    ['0xACFDA9Fd773C23c01f5d0CAE304CBEbE6b449677', 1],
+    ['0x68EBB057645258Cc62488fD198A0f0fa3FD6e8fb', 1],
+  ];
+
+  // This one was not added to the GaugeController, so we will not be using it. Adding it here for completion only.
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const optimismRootGauges = ['0x0256B02F3b19e71AF03B4e2A6731ca35106f5439'];
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:57]: We now have optimism gauges

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    ['0xF0ea3559Cf098455921d74173dA83fF2f6979495', 1],
+    ['0xB0de49429fBb80c635432bbAD0B3965b28560177', 1],
+    ['0x359EA8618c405023Fc4B98dAb1B01F373792a126', 1],
+    ['0xc77E5645Dbe48d54afC06655e39D3Fe17eB76C1c', 1],
+    ['0x899F737750db562b88c1E412eE1902980D3a4844', 1],
+    ['0x6cb1A77AB2e54d4560fda893E9c738ad770da0B0', 1],
+    ['0x6823DcA6D70061F2AE2AAA21661795A2294812bF', 1],
+    ['0xACFDA9Fd773C23c01f5d0CAE304CBEbE6b449677', 1],
+    ['0x68EBB057645258Cc62488fD198A0f0fa3FD6e8fb', 1],
+  ];
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:53]: Some of these might need updating

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+      const gaugesAmountAboveMinWeight = getGaugeDataAboveMinWeight(GaugeType.Arbitrum, minRelativeWeight).length;
+      const singleGaugeBridgeCost = await arbitrumGauge.getTotalBridgeCost();
+
+      // Bridge cost per gauge is always the same, so total cost is (single gauge cost) * (number of gauges).
+      expect(await L2GaugeCheckpointer.getTotalBridgeCost(minRelativeWeight)).to.be.almostEqual(
+        singleGaugeBridgeCost * gaugesAmountAboveMinWeight
+      );
+    });
+  }
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:229]: Move this fn to the context in which it is used

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+  function itChecksTotalBridgeCost(minRelativeWeight: BigNumber) {
+    it('checks total bridge cost', async () => {
+      const arbitrumGauge = await task.instanceAt('ArbitrumRootGauge', gauges.get(GaugeType.Arbitrum)![0].address);
+
+      const gaugesAmountAboveMinWeight = getGaugeDataAboveMinWeight(GaugeType.Arbitrum, minRelativeWeight).length;
+      const singleGaugeBridgeCost = await arbitrumGauge.getTotalBridgeCost();
+
+      // Bridge cost per gauge is always the same, so total cost is (single gauge cost) * (number of gauges).
+      expect(await L2GaugeCheckpointer.getTotalBridgeCost(minRelativeWeight)).to.be.almostEqual(
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:225]: Can't we test strict equality here? The values sould match exactly

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+      ];
+    });
+
+    afterEach('mark checkpointed gauges to consider in the next iteration', () => {
+      checkpointedGauges.push(...gaugeDataAboveMinWeight);
+    });
+
+    context('when threshold is 1', () => {
+      minRelativeWeight = fp(1);
+      itCheckpointsGaugesAboveRelativeWeight();
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:174]: I'd instead check here that this returns 0 (since no gauge fulfills this criteria). It's good to have a couple tests with hardcoded values and not depend too much on the 'smart' tests.

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    });
+  });
+
+  describe('checkpoint', () => {
+    const checkpointedGauges: GaugeData[] = [];
+    let gaugeDataAboveMinWeight: GaugeData[] = [];
+    let minRelativeWeight: BigNumber;
+
+    // Gauges won't be checkpointed twice, so when the threshold is lowered and more gauges get above the threshold
+    // we need to filter out those that have already been checkpointed.
```
Comment (nventuro) [inline pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts:160]: Isn't it easier to use snapshots? The test setup otherwise becomes quite fragile

File: pkg/deployments/README.md (hunk 1 above)
Comment (jubeira) [reply pkg/deployments/README.md:105]: Done.

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts (hunk 2 above)
Comment (jubeira) [reply pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts]: Updated to a block right before last week's manual checkpoints.

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts (hunk 4 above)
Comment (jubeira) [reply pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts]: Done; I've added the txs for last week's update too.

The list in the test matches the gauges checkpointed manually except for one, which is added for completion.

I double checked the rogue gauge, and it was probably skipped intentionally. Its relative weight for the 2 periods that it gets checkpointed in the test was 0.

File: pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts (hunk 8 above)
Comment (jubeira) [reply pkg/deployments/tasks/20221101-l2-gauge-checkpointer/test/task.fork.ts]: I found it a bit weird to do so back then in a fork test, but I think it makes sense.

I've added an empty `sharedBeforeEach` that takes the snapshot and it does its job; no need to keep track of the checkpointed gauges now.

**Comment by jubeira** [comment]
Comments addressed.
tl;dr; block # is right before last week's update. The automatic checkpoint matches the manual checkpoint except for one gauge with 0 weight that was skipped in the manual one, so it's looking good.

**Comment by jubeira** [comment]
It might make sense to supersede this with #1995 (similar fork tests / deployment task, but using the new adaptor entrypoint).

**Review by nventuro — APPROVED** [review]
Checked the build info again, this is good to go! The only thing I'd change is the name of the task, to more accurately reflect the actual deployment date.
